File: app/src/main/java/com/hover/stax/transfers/TransferViewModel.java
package com.hover.stax.transfers;

import android.app.Application;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.google.i18n.phonenumbers.NumberParseException;
import com.hover.sdk.actions.HoverAction;
import com.hover.stax.R;
import com.hover.stax.channels.Channel;
import com.hover.stax.contacts.PhoneHelper;
import com.hover.stax.contacts.StaxContact;
import com.hover.stax.requests.Request;
import com.hover.stax.schedules.Schedule;
import com.hover.stax.utils.DateUtils;
import com.hover.stax.utils.Utils;

import java.util.List;

public class TransferViewModel extends AbstractFormViewModel {
	final private String TAG = "TransferViewModel";

	private MutableLiveData<String> amount = new MutableLiveData<>();
	private MutableLiveData<StaxContact> contact = new MutableLiveData<>();
	private MutableLiveData<String> note = new MutableLiveData<>();

	protected LiveData<Request> request = new MutableLiveData<>();

	public TransferViewModel(Application application) {
		super(application);
	}

	void setType(String transaction_type) {
		type = transaction_type;
	}

	void setAmount(String a) {
		amount.postValue(a);
	}

	LiveData<String> getAmount() {
		if (amount == null) {
			amount = new MutableLiveData<>();
		}
		return amount;
	}

	void setContact(String contact_ids) {
		if (contact_ids == null) return;
		new Thread(() -> {
			List<StaxContact> contacts = repo.getContacts(contact_ids.split(","));
			if (contacts.size() > 0) contact.postValue(contacts.get(0));
		}).start();
	}

	void setContact(StaxContact c) {
		contact.setValue(c);
	}
	void forceUpdateContactUI() {contact.postValue(contact.getValue());}

	LiveData<StaxContact> getContact() {
		if (contact == null) { contact = new MutableLiveData<>(); }
		return contact;
	}

	void setRecipient(String r) {
		if (contact.getValue() != null && contact.getValue().toString().equals(r)) { return; }
		contact.setValue(new StaxContact(r));
	}

	void setRecipientSmartly(Request r, Channel channel) {
		if(channel !=null && r!=null) {
			try {
				String formattedPhone = PhoneHelper.getInternationalNumber(channel.countryAlpha2, r.requester_number);
				new Thread(() -> {
					StaxContact contactValue = repo.getContactByPhone(formattedPhone);
					if (contactValue == null) {
						//Check again without internationalizing number, in case the value is a bank account number;
						contactValue = repo.getContactByPhone(r.requester_number);
					}
					if (contactValue != null) contact.postValue(contactValue);
				}).start();
			} catch (NumberParseException e) {
				Utils.logErrorAndReportToFirebase(TAG, e.getMessage(), e);
			}
		}
	}

	public LiveData<Schedule> getSchedule() {
		if (schedule == null) { schedule = new MutableLiveData<>(); }
		return schedule;
	}

	public LiveData<Request> getRequest() {
		if (request == null) { request = new MutableLiveData<>(); }
		return request;
	}

	void setNote(String r) {
		note.postValue(r);
	}

	LiveData<String> getNote() {
		if (note == null) {
			note = new MutableLiveData<>();
			note.setValue(" ");
		}
		return note;
	}

	String amountErrors() {
		if (amount.getValue() != null && !amount.getValue().isEmpty() && amount.getValue().matches("[\\d.]+") && !amount.getValue().matches("[0]+"))
			return null;
		return getApplication().getString(R.string.amount_fielderror);
	}

	String recipientErrors(HoverAction a) {
		if (a != null && a.requiresRecipient() && contact.getValue() == null)
			return getApplication().getString(a.isPhoneBased() ? R.string.transfer_error_recipient_phone : R.string.transfer_error_recipient_account);
		return null;
	}

	public LiveData<Request> decrypt(String encryptedLink) {
		request = repo.decrypt(encryptedLink, getApplication());
		return request;
	}

	public void view(Schedule s) {
		schedule.setValue(s);
		setType(s.type);
		setAmount(s.amount);
		setContact(s.recipient_ids);
		setNote(s.note);
	}

	public void view(Request r) {
		setAmount(r.amount);
		setContact(r.requestee_ids);
		setNote(r.note);
	}

	public void checkSchedule() {
		if (schedule.getValue() != null) {
			Schedule s = schedule.getValue();
			if (s.end_date <= DateUtils.today()) {
				s.complete = true;
				repo.update(s);
			}
		}
	}

	public void saveContact() {
		final StaxContact c = contact.getValue();
		if (c != null) {
			new Thread(() -> {
				c.lastUsedTimestamp = DateUtils.now();
				repo.save(c);
			}).start();
		}
	}
}

File: app/src/main/java/com/hover/stax/transfers/TransferViewModel.kt
/*
 * Copyright 2022 Stax
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.hover.stax.transfers

import android.app.Application
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import com.hover.sdk.actions.HoverAction
import com.hover.stax.R
import com.hover.stax.data.contact.ContactRepo
import com.hover.stax.database.models.PhoneHelper
import com.hover.stax.database.models.StaxContact
import com.hover.stax.database.models.Request
import com.hover.stax.data.requests.RequestRepo
import com.hover.stax.data.schedule.ScheduleRepo
import com.hover.stax.core.AnalyticsUtil
import com.hover.stax.core.DateUtils
import com.yariksoffice.lingver.Lingver
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import timber.log.Timber

class TransferViewModel(
    application: Application,
    private val requestRepo: RequestRepo,
    contactRepo: ContactRepo,
    scheduleRepo: ScheduleRepo
) : AbstractFormViewModel(application, contactRepo, scheduleRepo) {

    val amount = MutableLiveData<String?>()
    val contact = MutableLiveData<StaxContact?>()
    val note = MutableLiveData<String?>()

    val isLoading = MutableLiveData(false)

    fun setAmount(a: String?) = amount.postValue(a)

    fun setContact(contactId: String) = viewModelScope.launch(Dispatchers.IO) {
        contact.postValue(contactRepo.getContact(contactId))
    }

    fun setContact(sc: StaxContact?) = sc?.let {
        contact.postValue(it)
    }

    fun setRecipientNumber(str: String) {
        if (contact.value != null && contact.value.toString() == str) return
        if (str.isEmpty()) {
            contact.value =
                StaxContact()
        } else {
            contact.value = StaxContact(str)
        }
    }

    private fun setRecipientSmartly(r: Request?, countryAlpha2: String?) =
        viewModelScope.launch(Dispatchers.IO) {
            r?.let {
                try {
                    val formattedPhone = PhoneHelper.getNationalSignificantNumber(
                        it.requester_number!!,
                        countryAlpha2 ?: Lingver.getInstance().getLocale().country
                    )
                    val sc = contactRepo.getContactByPhone(formattedPhone)
                    contact.postValue(
                        sc ?: StaxContact(
                            r.requester_number
                        )
                    )
                    isLoading.postValue(false)
                } catch (e: NumberFormatException) {
                    com.hover.stax.core.AnalyticsUtil.logErrorAndReportToFirebase(
                        TransferViewModel::class.java.simpleName, e.message!!, e
                    )
                }
            }
        }

    private fun setNote(n: String?) = note.postValue(n)

    fun amountErrors(a: HoverAction?): String? {
        val regex = a?.getStepByVar(HoverAction.AMOUNT_KEY)?.optString("valid_response_regex")
        return if ((amount.value.isNullOrEmpty() || !amount.value!!.matches("[\\d.]+".toRegex()) || amount.value!!.matches("[0]+".toRegex())) ||
            (!regex.isNullOrEmpty() && !amount.value!!.matches(regex.toRegex()))
        ) getString(R.string.amount_fielderror)
        else null
    }

    fun recipientErrors(a: HoverAction?): String? {
        return when {
            (a != null && a.requiresRecipient() && (contact.value == null || contact.value?.accountNumber.isNullOrEmpty())) -> getString(if (a.isPhoneBased) R.string.transfer_error_recipient_phone else R.string.transfer_error_recipient_account)
            else -> null
        }
    }

    fun wrapExtras(action: HoverAction): HashMap<String, String> {
        val extras: HashMap<String, String> = hashMapOf()
        if (amount.value != null) extras[HoverAction.AMOUNT_KEY] = amount.value!!
        if (contact.value != null && contact.value?.accountNumber != null) {
            extras[StaxContact.ID_KEY] = contact.value!!.id
            extras[HoverAction.PHONE_KEY] = contact.value!!.accountNumber
            extras[HoverAction.ACCOUNT_KEY] = generateRecipientAccount(action)
        }
        if (note.value != null) extras[HoverAction.NOTE_KEY] = note.value!!
        return extras
    }

    private fun generateRecipientAccount(action: HoverAction): String {
        return if (action.bonus_percent > 0 && action.getStepByVar("account").has("prefix")) {
            val prefix = action.getStepByVar("account").optString("prefix")
            if (action.isPhoneBased)
                prefix + PhoneHelper.normalizeNumberByCountry(
                    contact.value!!.accountNumber, action.country_alpha2, action.to_country_alpha2
                )
            else prefix + contact.value!!.accountNumber
        } else contact.value!!.accountNumber
    }

    fun load(encryptedString: String) = viewModelScope.launch {
        isLoading.postValue(true)
        val r: Request? = requestRepo.decrypt(encryptedString, getApplication())
        Timber.v("Loaded request %s", r)
        r?.let {
            setRecipientSmartly(r, r.requester_country_alpha2)
            setAmount(r.amount)
            setNote(r.note)
            com.hover.stax.core.AnalyticsUtil.logAnalyticsEvent(getString(R.string.loaded_request_link), getApplication())
        }
    }

    fun saveContact() {
        viewModelScope.launch(Dispatchers.IO) {
            contact.value?.let { sc ->
                val lookup = StaxContact.getContactByPhoneValue(sc.accountNumber, "", contactRepo)
                val c = if (!sc.hasName() && lookup != null) {
                    lookup
                } else {
                    sc
                }
                c.lastUsedTimestamp = com.hover.stax.core.DateUtils.now()
                contactRepo.save(c)
                contact.postValue(c)
            }
        }
    }

    override fun reset() {
        super.reset()
        amount.value = null
        contact.value = null
        note.value = null
    }
}